### PR TITLE
GTC Optimizations Part 2: Loop Merging, Caches, On-The-Fly Fusion, Improved Temporary Replacement

### PR DESCRIPTION
**Pull request by fthaler**
## Description

Enables the following optimization passes in the GTC backend:
- AdjacentLoopMerging (merges contiguous vertical loops into a single one)
- LocalTemporariesToScalars (replaces temporaries used in a single horizontal execution by local scalar variables)
- WriteBeforeReadTemporariesToScalars (replaces temporaries that are written before read in each horizontal exec. by local scalars)
- OnTheFlyMerging (transforms dependent horizontal executions to redundant on-the-fly computations)
- NoFieldAccessPruning (prunes computations without field accesses and thus without effect)
- IJCacheDetection (detects computations where ij-caches can be applied)
- KCacheDetection (detects loops where k-caches can be applied)
- PruneKCacheFills (prunes unnecessary k-cache fills)
- PruneKCacheFills (prunes unnecessary k-cache flushes)

Further, there is FillFlushToLocalKCaches which converts filling or flushing k-caches to local k-caches with additional fill and flush statements. This is not used in the GT-based backends yet but is required for the coming GTC-CUDA backend.

## Requirements

Before submitting this PR, please make sure:

- [x] The code builds cleanly without new errors or warnings
- [x] The code passes all the existing tests
- [x] If this PR adds a new feature, new tests have been added to test these
new features
- [x] All relevant documentation has been updated or added

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

**Comment by fthaler** [comment]
GTBench convergence tests pass.

File: src/gtc/passes/oir_optimizations/horizontal_execution_merging.py [hunk 1]
```diff
@@ -100,3 +79,140 @@ def visit_VerticalLoopSection(
...
+        **kwargs: Any,
+    ) -> Union[oir.FieldAccess, oir.ScalarAccess]:
+        if offset_symbol_map:
+            offset = self.visit(node.offset, **kwargs)
+            key = node.name, (offset.i, offset.j, offset.k)
+            if key in offset_symbol_map:
+                return oir.ScalarAccess(name=offset_symbol_map[key], dtype=node.dtype)
+        return self.generic_visit(node, **kwargs)
+
+    def _merge(
```
Comment (havogt) [inline src/gtc/passes/oir_optimizations/horizontal_execution_merging.py:113]: note to myself: play with some examples

File: src/gtc/passes/oir_optimizations/temporaries.py [hunk 2]
```diff
@@ -93,11 +118,32 @@ def visit_Stencil(self, node: oir.Stencil, **kwargs: Any) -> oir.Stencil:
...
-            params=node.params,
-            vertical_loops=self.visit(
-                node.vertical_loops, local_tmps=local_tmps, symtable=node.symtable_, **kwargs
-            ),
-            declarations=[d for d in node.declarations if d.name not in local_tmps],
-        )
+        return super().visit_Stencil(node, tmps_to_replace=local_tmps)
+
+
+class WriteBeforeReadTemporariesToScalars(TemporariesToScalarsBase):
```
Comment (havogt) [inline src/gtc/passes/oir_optimizations/temporaries.py:124]: note to myself: try to break it with `If`s and intervals.

File: src/gtc/passes/oir_optimizations/vertical_loop_merging.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            b_lim = b_interval.start
+        return a_lim.level == b_lim.level and a_lim.offset == b_lim.offset
+
+    @staticmethod
+    def _merge(a: oir.VerticalLoop, b: oir.VerticalLoop) -> oir.VerticalLoop:
+        sections = a.sections + b.sections
+        return oir.VerticalLoop(
+            loop_order=a.loop_order,
+            sections=sections,
+            caches=[],  # TODO: add support for cache merging?
```
Comment (havogt) [inline src/gtc/passes/oir_optimizations/vertical_loop_merging.py:44]: regarding the TODO: probably the merge should run before the cache optimizations? If yes, then we could warn or assert here if caches in a.caches and b.caches are not empty.

**Review by havogt — CHANGES_REQUESTED** [review]
This is a partial review. I didn't look at the tests. Will do after the factory_boy PR is merged. Maybe we can also talk a bit tomorrow: you explain the algorithms to me and we discuss corner cases.

**Comment by havogt** [comment]
```python
def demo_copy(in_field: gtscript.Field[dtype], out_field: gtscript.Field[dtype]):
    with computation(FORWARD):
        with interval(...):
            tmp = 0
    with computation(FORWARD):
        with interval(0,1):
            # tmp = 0
            out_field = 0
        with interval(1,None):
            tmp = in_field  # noqa: F841 [assigned but not used]
            out_field = tmp[0,0,0]
```
Triggers compilation error

File: src/gtc/passes/oir_optimizations/horizontal_execution_merging.py [hunk 4]
```diff
@@ -100,3 +79,140 @@ def visit_VerticalLoopSection(
...
+        new_symbol_name: Callable[[str], str],
+    ) -> List[oir.HorizontalExecution]:
+        """Recursively merge horizontal executions.
+
+        Uses the following algorithm:
+        1. Get output fields of the first horizontal execution.
+        2. Check in which following h. execs. the outputs are read.
+        3. Duplicate the body of the first h. exec. for each read access (with corresponding offset) and prepend it to the depending h. execs.
+        4. Recurse on the resulting h. execs.
+        """
```
Comment (havogt) [inline src/gtc/passes/oir_optimizations/horizontal_execution_merging.py:127]: describe limitations:
- copies whole h exec, even if only a few statements would be relevant
- heuristic would be needed for size and for heavy functions (e.g. `exp`)

File: src/gtc/passes/oir_optimizations/vertical_loop_merging.py (hunk 3 above)
Comment (fthaler) [reply src/gtc/passes/oir_optimizations/vertical_loop_merging.py:44]: Replaced the TODO by a warning in case there are removed caches. The pass should indeed run before cache detection and that’s also what’s implemented in the current GTC backends.

File: src/gtc/passes/oir_optimizations/horizontal_execution_merging.py (hunk 4 above)
Comment (fthaler) [reply src/gtc/passes/oir_optimizations/horizontal_execution_merging.py:137]: Added comment and configurable merging limitations (body size and expensive functions).

**Comment by fthaler** [comment]
> ```python
> def demo_copy(in_field: gtscript.Field[dtype], out_field: gtscript.Field[dtype]):
>     with computation(FORWARD):
>         with interval(...):
>             tmp = 0
>     with computation(FORWARD):
>         with interval(0,1):
>             # tmp = 0
>             out_field = 0
>         with interval(1,None):
>             tmp = in_field  # noqa: F841 [assigned but not used]
>             out_field = tmp[0,0,0]
> ```
> 
> Triggers compilation error

Fixed with additional NoFieldAccessPruning pass that prunes all horizontal executions (and parent loops) without any field accesses.

File: src/gtc/passes/oir_optimizations/caches.py [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+        filling_fields: Dict[str, str] = {
+            c.name: new_symbol_name(c.name)
+            for c in node.caches
+            if isinstance(c, oir.KCache) and c.fill
+        }
+        flushing_fields: Dict[str, str] = {
+            c.name: filling_fields[c.name] if c.name in filling_fields else new_symbol_name(c.name)
+            for c in node.caches
+            if isinstance(c, oir.KCache) and c.flush
+        }
```
Comment (egparedes) [inline src/gtc/passes/oir_optimizations/caches.py:450]: Not important, but I think a more explicit for-loop would be more readable here:

```suggestion
        filling_fields: Dict[str, str] = {}
        flushing_fields: Dict[str, str] = {}
        for c in node.caches:
            if isinstance(c, oir.KCache):
                if c.fill:
                    filling_fields[c.name] = new_symbol_name(c.name)
                if c.flush:
                    flushing_fields[c.name] = filling_fields.get(c.name, new_symbol_name(c.name))
    }
```

**Review by egparedes — APPROVED** [review]
It looks good to me. Just added a few comments as feedback, but nothing that blocks merging this PR.

**Comment by egparedes** [comment]
bors try

**Comment by fthaler** [comment]
bors try

**Comment by fthaler** [comment]
bors try

**Comment by fthaler** [comment]
bors try

**Comment by havogt** [comment]
bors try
